### PR TITLE
IOS: implement /dev/aes & /dev/sha

### PR DESCRIPTION
**Pull request by DacoTaco**
with libogc supporting the sha & aes engine, and applications starting to use them, its become needed for dolphin to get a basic implementation of these devices.

using mbedtls i have implemented the devices, and they are confirmed working.

you can test them using the following dol : [crypto.zip](https://github.com/dolphin-emu/dolphin/files/12937862/crypto.zip)
this is an example i wrote for libogc

the implementation is done by looking at the work done for StarStruck

### CODE REVIEW

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
The new files need to be added to CMakeLists.txt.

**Comment by DacoTaco** [comment]
> The new files need to be added to CMakeLists.txt.

which one? im not familiar with CMakeLists, nor did i find the DI stuff anywhere (which is what i based my edits on)

File: Source/Core/DolphinLib.vcxproj
Comment (Pokechu22) [inline Source/Core/DolphinLib.vcxproj:1]: I don't think any of the changes in this file are needed - I'm guessing they were automatically produced by Visual Studio?

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2008 Dolphin Emulator Project
```
Comment (Pokechu22) [inline Source/Core/Core/IOS/Crypto/AesDevice.cpp:1]: This should probably be 2023, same with the other files.

**Review by Pokechu22 — COMMENTED** [review]
Do we need to do anything with the magic 0x10000 / 0x10001 / 0x10002 FDs mentioned on wiibrew's articles on [/dev/aes](https://wiibrew.org/wiki//dev/aes) and [/dev/sha](https://wiibrew.org/wiki//dev/sha)? Or are those only relevant to IOS and not relevant externally?

(I haven't looked at the implementation at all.)

File: Source/Core/DolphinLib.vcxproj
Comment (DacoTaco) [reply Source/Core/DolphinLib.vcxproj:1]: they were, and i reverted them :)

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp (hunk 1 above)
Comment (DacoTaco) [reply Source/Core/Core/IOS/Crypto/AesDevice.cpp:1]: copy paste fail to the rescue :p

**Comment by DacoTaco** [comment]
> Do we need to do anything with the magic 0x10000 / 0x10001 / 0x10002 FDs mentioned on wiibrew's articles on [/dev/aes](https://wiibrew.org/wiki//dev/aes) and [/dev/sha](https://wiibrew.org/wiki//dev/sha)? Or are those only relevant to IOS and not relevant externally?
> 
> (I haven't looked at the implementation at all.)

i think to be correct, they should be implemented. however, i don't know if thats for all ios versions, nor where those come from (they weren't found when i was looking at the aes/sha modules tbh)

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "Core/IOS/Crypto/AesDevice.h"
```
Comment (JosJuice) [inline Source/Core/Core/IOS/Crypto/AesDevice.cpp:3]:
```suggestion
// SPDX-License-Identifier: GPL-2.0-or-later

#include "Core/IOS/Crypto/AesDevice.h"
```

**Comment by DacoTaco** [comment]
@Pokechu22 : any opinions if i were to add the fd's like this : 
``` 
  std::shared_ptr<Device> dev = std::make_shared<AesDevice>(*this, "/dev/aes");
  m_fdmap[0x10000] = dev;
  AddDevice(std::make_unique<AesDevice>(&dev));

  std::shared_ptr<Device> dev = std::make_shared<ShaDevice>(*this, "/dev/sha");
  m_fdmap[0x10001] = dev;
  m_fdmap[0x10002] = dev;
  AddDevice(std::make_unique<ShaDevice>(&dev));
```

or is there some code that already exists for this?

**Comment by Pokechu22** [comment]
I don't think there's any existing code that does it; I'm pretty sure these are the only "magic" FDs that exist. That's probably reasonable (though I wonder if it's possible to close those magic FDs).

File: Source/Core/Core/IOS/Crypto/Sha.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+{
+  std::copy(std::begin(src.Length), std::end(src.Length), std::begin(dest.total));
+  std::copy(std::begin(src.States), std::end(src.States), std::begin(dest.state));
+}
+
+void ConvertContext(mbedtls_sha1_context& src, ShaDevice::ShaContext& dest)
+{
+  std::copy(std::begin(src.total), std::end(src.total), std::begin(dest.Length));
+  std::copy(std::begin(src.state), std::end(src.state), std::begin(dest.States));
+}
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/Sha.cpp:39]: These functions shouldn't be visible from outside the compilation unit so they should be `static`. `src` should be `const`. Also consider making `dest` a pointer so the direction of the conversion is more obvious at the call site.

**Comment by DacoTaco** [comment]
> I don't think there's any existing code that does it; I'm pretty sure these are the only "magic" FDs that exist. That's probably reasonable (though I wonder if it's possible to close those magic FDs).

@Pokechu22 : looking back at starstruck, it seems a ios_close to those fd's results in a ios_close being sent, but the fd not being cleared. i'll make dolphin respond the same way

File: Source/Core/Core/IOS/Crypto/Sha.cpp (hunk 3 above)
Comment (Pokechu22) [reply Source/Core/Core/IOS/Crypto/Sha.cpp:39]: Yes, that's what [our style guide](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md#functions) would want. (We don't recommend raw pointers created using `new`, but a raw pointer as a function parameter where there aren't any ownership considerations is fine).

File: Source/Core/Core/IOS/Crypto/Sha.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "Core/IOS/Crypto/Sha.h"
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/Sha.cpp:3]: Please add an empty line here.

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  }
+  case AesIoctlv::Decrypt:
+  case AesIoctlv::Encrypt:
+  {
+    if (!request.HasNumberOfValidVectors(2, 2))
+      break;
+
+    if (request.in_vectors[1].size != 0x10 || (request.in_vectors[1].address & 3) != 0 ||
+        request.io_vectors[1].size != 0x10 || (request.io_vectors[1].address & 3) != 0)
+      break;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/AesDevice.cpp:58]: The if is multi-line so this should have braces.

**Review by AdmiralCurtiss — APPROVED** [review]
Very minor style nits but yeah this looks good to me otherwise. I'll merge this later today or tomorrow once those are resolved if no one else has any objections.

File: Source/Core/Core/IOS/Crypto/Sha.cpp (hunk 4 above)
Comment (DacoTaco) [reply Source/Core/Core/IOS/Crypto/Sha.cpp]: adjusted in new force push

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp (hunk 5 above)
Comment (DacoTaco) [reply Source/Core/Core/IOS/Crypto/AesDevice.cpp]: adjusted in new force push

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+std::optional<IPCReply> AesDevice::Open(const OpenRequest& request)
+{
+  return Device::Open(request);
+}
+
+std::optional<IPCReply> AesDevice::IOCtlV(const IOCtlVRequest& request)
+{
+  auto& system = GetSystem();
+  auto& memory = system.GetMemory();
+  HLE::ReturnCode returnCode = IPC_EINVAL;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/AesDevice.cpp:34]: `returnCode` -> `return_code`

File: Source/Core/Core/IOS/Crypto/Sha.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/Crypto/Sha.h"
+
+#include <algorithm>
+#include <iterator>
+#include <mbedtls/sha1.h>
+#include <memory>
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/Sha.cpp:9]: This also wants `#include <optional>`, `#include <vector>`, `#include <array>`.

File: Source/Core/Core/IOS/Crypto/Sha.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/Sha.h:5]: `#include <array>`

File: Source/Core/Core/IOS/Crypto/Sha.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    InitState = 0x00,
+    ContributeState = 0x01,
+    FinalizeState = 0x02,
+    ShaCommandUnknown = 0x0F
+  };
+
+  struct ShaContext
+  {
+    std::array<u32, 5> States;
+    std::array<u32, 2> Length;  // length in bits of total data contributed to SHA-1 hash
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/Sha.h:29]: These should be lowercase `states` and `length`.

File: Source/Core/Core/IOS/Crypto/Sha.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  if (!ret && command == ShaIoctlv::FinalizeState)
+    memory.CopyToEmu(request.io_vectors[1].address, output_hash.data(), output_hash.size());
+
+  mbedtls_sha1_free(&context);
+  return ret ? HLE::ReturnCode::IPC_EACCES : HLE::ReturnCode::IPC_SUCCESS;
+}
+
+std::optional<IPCReply> ShaDevice::IOCtlV(const IOCtlVRequest& request)
+{
+  HLE::ReturnCode returnCode = IPC_EINVAL;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/Crypto/Sha.cpp:80]: -> `return_code`

**Review by AdmiralCurtiss — COMMENTED** [review]
A few more things... sorry for not noticing those before.

**Comment by DacoTaco** [comment]
> A few more things... sorry for not noticing those before.

no problem, thats what happens when somebody who doesn't know the codebase/workflows/style all that much contributes hehe :')

File: Source/Core/Core/IOS/Crypto/AesDevice.cpp [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    if (request.in_vectors[1].size != 0x10 || (request.in_vectors[1].address & 3) != 0 ||
+        request.io_vectors[1].size != 0x10 || (request.io_vectors[1].address & 3) != 0)
+    {
+      break;
+    }
+
+    std::vector<u8> input = std::vector<u8>(request.in_vectors[0].size);
+    std::vector<u8> output = std::vector<u8>(request.io_vectors[0].size);
+    std::vector<u8> key = std::vector<u8>(0x10);
+    std::vector<u8> iv = std::vector<u8>(0x10);
```
[the remark this answers is not in the document]
Comment (DacoTaco) [reply Source/Core/Core/IOS/Crypto/AesDevice.cpp]: should all have been resolved now :)

**Pull request merged**